File: src/wrappers/Waifu2xCaffe.java
package wrappers;/**
 * A rather peculiar class. So Waifu2x-Caffee lacks two stuff that my version of Dandere.lua (my modified waifu2x.lua)
 * file has: 1) ability to scale files in a specific order, 2)  and ability to scale files that don't exist yet.(it should
 * wait until the file exists).
 * <p>
 * 1) To combat the first, waifu2x-caffee can read files in lexiographic order, hence why the file names
 * to be fed into Waifu2xCaffee are in lexigraphic orderings.
 * <p>
 * 2) Waiuf2x-caffee can only scale what's currently in a folder, not files that don't exist yet. Furtheremore,
 * it doesn't know what images have been upscaled already, and will reupscale an image if it's already in the ouput folder.
 * The way we get around this is by having the 'output' folder become a buffer in a sense, and then signaling waifu2x-caffee
 * to upscale all the images in the 'output' folder, as it can only upscale a folder, not a list of files.
 * Once it's all upscaled, do a quick check of what got upscaled, delete them from the outputs Folder,
 * then signal the 'upscaleFolder' command in waifu2x-caffee to start the process again, upscaling all the images
 * in the 'outputs' folder at a given time.
 **/

import dandere2x.Utilities.DThread;
import dandere2x.Utilities.DandereUtils;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.PrintStream;
import java.util.ArrayList;

import static java.io.File.separator;

public class Waifu2xCaffe extends DThread {

    //debugging main
//    public static void main(String[] args){
//        String waifu2xCaffeCUIDir = "C:\\Users\\windwoz\\Desktop\\waifu2x-caffe\\waifu2x-caffe-cui.exe";
//        String outputsDir = "C:\\Users\\windwoz\\Desktop\\autodeletetest\\outputs";
//        String upscaledDir = "C:\\Users\\windwoz\\Desktop\\autodeletetest\\upscaled";
//        int count = 137;
//        String setting = "cudnn";
//
//        wrappers.Waifu2xCaffe sesh = new wrappers.Waifu2xCaffe(waifu2xCaffeCUIDir,outputsDir,upscaledDir,count,setting);
//        sesh.upscale();
//    }

    private int lexiConstant = 6;
    private String waifu2xCaffeDir;
    private String outputDir;
    private String upscaledDir;
    private String setting;
    private String noiseLevel;
    private ArrayList<Integer> upscaledFrames;
    private int frameCount;
    private Process waifu2xProc = null;
    private PrintStream log;
    private int scaledCount = 0;
    private double scaleFactor;

    public Waifu2xCaffe(String workspace, String waifu2xCaffeDir, String outputDir, String upscaledDir, int frameCount,
                        String setting, String noiseLevel, double scaleFactor, boolean isResume) {

        super(isResume);
        this.waifu2xCaffeDir = waifu2xCaffeDir;
        this.outputDir = outputDir;
        this.upscaledDir = upscaledDir;
        this.frameCount = frameCount;
        this.setting = setting;
        this.scaleFactor = scaleFactor;
        this.noiseLevel = noiseLevel;
        upscaledFrames = new ArrayList<Integer>();

        for (int x = 1; x < frameCount - 1; x++) {
            upscaledFrames.add(x);
        }

        if (isResume)
            resumeCondition();

        try {
            log = new PrintStream(new File(workspace + "logs" + separator + "waifu2x_logfile.txt"));
        } catch (FileNotFoundException e) {
            System.out.println("Fatal Error: Could not create file at " + workspace + "logs" + separator + "waifu2x_logfile.txt");
        }

    }

    //a function that exists soly for upscaling a single file.
    public static void upscaleFile(String waifu2xCaffeDir, String input, String output, String setting, String noiseLevel, double scaleFactor) {
        Process proc = null;
        Runtime run = Runtime.getRuntime();

        String upscaleCommand = waifu2xCaffeDir + " -i " + input + " -p " + setting + " -n " + noiseLevel + " -s " + scaleFactor + " -o " + output;
        try {
            proc = run.exec(upscaleCommand);
        } catch (IOException e) {
            e.printStackTrace();
        }

        synchronized (proc) {
            while (proc.isAlive()) {
            }
        }
    }

    @Override
    public void resumeCondition() {
//        System.out.println("resume condition");
//
//        for(int x = 0; x < upscaledFrames.size(); x++){
//            File temp = new File(upscaledDir + "output_" + DandereUtils.getLexiconValue(lexiConstant, x) + ".png");
//            File deleteFile = new File(outputDir + "output_" + DandereUtils.getLexiconValue(lexiConstant, x) + ".jpg");
//
//            if(temp.exists() && deleteFile.exists()){
//                System.out.println("Deleting");
//                deleteFile.delete();
//            }
//            System.out.println(temp.exists() + " " + deleteFile.exists());
//        }
    }

    @Override
    public void run() {

        System.out.println("waifu2x starting");
        shutdownHook();
        String upscaleCommand = waifu2xCaffeDir + " -i " + outputDir + " -p" + setting + " -n " + noiseLevel + " -s " + scaleFactor + " -o " + upscaledDir;


        /**
         * We keep track of how many frames are removed with 'scaledCount', and exit when we're done.
         */

        //TODO resuming waifu2xCaffee run not working properly.
        //files not being removed when they should from upscaled folder.
        while (upscaledFrames.size() > 1) {
            log.println(upscaledFrames.size());
            log.println(upscaledFrames);
            /**
             * Waifu2x-Caffee will niavely upscale everything in a folder. To mitigate this, we upscale images
             * in batches, in which at a given moment, create a list of all the images in 'outputs', upscale them
             * (the previous command), and now here we delete the ones that upscaled, so the new batch
             * of upscaled images won't be a part of the next batch.
             */


            //start backwards so .remove doesn't affect the ordering 
            for (int x = upscaledFrames.size() - 1; x >= 0; x--) {
                File temp = new File(upscaledDir + "output_" + DandereUtils.getLexiconValue(lexiConstant, upscaledFrames.get(x)) + ".png");
                File deleteFile = new File(outputDir + "output_" + DandereUtils.getLexiconValue(lexiConstant, upscaledFrames.get(x)) + ".jpg");


                if (temp.exists() && deleteFile.exists()) {
                    deleteFile.delete();
                    upscaledFrames.remove(upscaledFrames.get(x));
                } else if (temp.exists()) {
                    upscaledFrames.remove(upscaledFrames.get(x));
                }
            }

            Runtime run = Runtime.getRuntime();
            try {
                log.println("running command " + upscaleCommand);
                waifu2xProc = run.exec(upscaleCommand);
            } catch (IOException e) {
                e.printStackTrace();
            }

            /**
             * While waifu2x-caffee is upscaling the current batch of images, keep waiting
             */
            synchronized (waifu2xProc) {
                while (waifu2xProc.isAlive()) {
                }
            }

        }
    }

    //todo, test this function

    //if program exits and dandere2xCppProc is still running, close that up
    private void shutdownHook() {
        Runtime.getRuntime().addShutdownHook(new Thread() {
            public void run() {
                if (waifu2xProc.isAlive()) {
                    System.out.println("Unexpected shutdown! Waifu2x Process is alive!");
                    waifu2xProc.destroyForcibly();
                    System.out.println("Exiting..");
                }
            }
        });
    }


}

File: src/wrappers/waifu2x_caffe.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Name: Dandere2X waifu2x-caffe
Author: CardinalPanda
Date Created: March 22, 2019
Last Modified: April 2, 2019
"""
from dandere2x_core.dandere2x_utils import get_lexicon_value
import logging
import os
import subprocess
import threading
from context import Context


# temporary implementation of waifu2x-caffe wrapper
# note to self - add listener to delete files in real time(maybe?) for resume.
# Not sure if Video2x wants that as a feature, though.

# workspace, frame_count, waifu2x_caffe_dir, model_dir, output_dir, upscaled_dir, p_setting,
#                  noise_level, scale_factor

class Waifu2xCaffe(threading.Thread):
    def __init__(self, context: Context):
        self.frame_count = context.frame_count
        self.waifu2x_caffe_cui_dir = context.waifu2x_caffe_cui_dir
        self.differences_dir = context.differences_dir
        self.upscaled_dir = context.upscaled_dir
        self.process_type = context.process_type
        self.noise_level = context.noise_level
        self.scale_factor = context.scale_factor
        self.model_dir = context.model_dir
        self.workspace = context.workspace
        self.gpu_number = context.gpu_number

        threading.Thread.__init__(self)
        logging.basicConfig(filename=self.workspace + 'waifu2x.log', level=logging.INFO)

    @staticmethod
    def upscale_file(context: Context, input_file : str, output_file: str):
        # load variables from context
        process_type = context.process_type
        noise_level = context.noise_level
        scale_factor = context.scale_factor
        model_dir = context.model_dir
        waifu2x_caffe_cui_dir = context.waifu2x_caffe_cui_dir
        gpu_number = context.gpu_number

        logger = logging.getLogger(__name__)

        exec = [waifu2x_caffe_cui_dir,
                "-i", input_file,
                "-p", process_type,
                "-n", noise_level,
                "-s", scale_factor,
                "--gpu", gpu_number,
                "-o", output_file]

        # if the user is using a custom model
        if model_dir != "default":
            exec.append("--model_dir")
            exec.append(model_dir)

        logger.info("manually upscaling file")
        logger.info(exec)
        subprocess.run(exec)

    # The current Dandere2x implementation requires files to be removed from the folder
    # During runtime. As files produced by Dandere2x don't all exist during the initial
    # Waifu2x call, various work arounds are in place to allow Dandere2x and Waifu2x to work in real time.

    # Briefly, 1) Create a list of names that will be upscaled by waifu2x,
    #          2) Call waifu2x to upscale whatever images are in 'differences' folder
    #          3) After waifu2x call is finished, delete whatever files were upscaled, and remove those names from list.
    #             (this is to prevent Waifu2x from re-upscaling the same image again)
    #          4) Repeat this process until all the names are removed.
    def run(self):
        logger = logging.getLogger(__name__)

        exec = [self.waifu2x_caffe_cui_dir,
                "-i", self.differences_dir,
                "-p", self.process_type,
                "-n", self.noise_level,
                "-s", self.scale_factor,
                "--gpu", self.gpu_number,
                "-o", self.upscaled_dir]

        if self.model_dir != "default":
            exec.append("--model_dir")
            exec.append(self.model_dir)

        logger.info("waifu2xcaffe session")
        logger.info(exec)

        # make a list of names that will eventually (past or future) be upscaled
        names = []
        for x in range(1, self.frame_count):
            names.append("output_" + get_lexicon_value(6, x) + ".png")

        count_removed = 0

        # remove from the list images that have already been upscaled
        for item in names[::-1]:
            if os.path.isfile(self.upscaled_dir + item):
                names.remove(item)
                count_removed += 1

        if count_removed:
            logger.info("Already have " + str(count_removed) + " upscaled")

        # while there are pictures that have yet to be upscaled, keep calling the upscale command
        while names:
            logger.info("Frames remaining before batch: ")
            logger.info(len(names))
            subprocess.run(exec, stdout=open(os.devnull, 'wb'))
            for item in names[::-1]:
                if os.path.isfile(self.upscaled_dir + item):
                    os.remove(self.differences_dir + item)
                    names.remove(item)
